src/java.base/share/classes/sun/security/provider/SecureRandom.java

```diff
@@ -218,53 +274,56 @@ private static void updateState(byte[] state, byte[] output) {
      * @param result the array to be filled in with random bytes.
      */
     @Override
-    public synchronized void engineNextBytes(byte[] result) {
-        int index = 0;
-        int todo;
-        byte[] output = remainder;
-
-        if (state == null) {
-            byte[] seed = new byte[DIGEST_SIZE];
-            SeederHolder.seeder.engineNextBytes(seed);
-            state = digest.digest(seed);
-        }
+    public void engineNextBytes(byte[] result) {
+        objLock.lock();
```

Comment:
A minor drawback, the keyword in the method declaration makes the synchronization more explicit. An alternative is to retain the synchronization the on object monitor by e.g.

```
private boolean blocked = false;

public synchronized void engineNextBytes {
   while (blocked) {
     this.wait()
   }
   ...
}

public synchronized void beforeCheckpoint {
  blocked = true
  ...
}
public synchronized void afterRestore {
  ...
  blocked = false;
  this.notifyAll();
}
```

However, a dedicated ReentrantLock makes the implementation a bit shorter. So, up to you.

Reply:
Yes, it makes sense. But using Lock object in the beforeCheckpoint() -> afterRestore() methods looks more natural.
Also, in your case, we'll have to check for interruptedException and modify engineSetSeed() accordingly.
If there are no more objections, I would use objLock for synchronization.